homeassistant/components/select/__init__.py

```diff
@@ -0,0 +1,96 @@
+"""Component to allow selecting an option from a list as platforms."""
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+from typing import Any, cast, final
+
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.config_validation import (  # noqa: F401
+    PLATFORM_SCHEMA,
+    PLATFORM_SCHEMA_BASE,
+)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.typing import ConfigType
+
+from .const import ATTR_OPTION, ATTR_OPTIONS, DOMAIN, SERVICE_SELECT_OPTION
+
+SCAN_INTERVAL = timedelta(seconds=30)
+
+ENTITY_ID_FORMAT = DOMAIN + ".{}"
+
+MIN_TIME_BETWEEN_SCANS = timedelta(seconds=10)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up Select entities."""
+    component = hass.data[DOMAIN] = EntityComponent(
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+    await component.async_setup(config)
+
+    component.async_register_entity_service(
+        SERVICE_SELECT_OPTION,
+        {vol.Required(ATTR_OPTION): cv.string},
+        "async_select_option",
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    return cast(bool, await hass.data[DOMAIN].async_setup_entry(entry))
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    return cast(bool, await hass.data[DOMAIN].async_unload_entry(entry))
+
+
+class SelectEntity(Entity):
+    """Representation of a Select entity."""
+
+    _attr_current_option: str | None
+    _attr_options: list[str]
+    _attr_state: None = None
+
+    @property
+    def capability_attributes(self) -> dict[str, Any]:
+        """Return capability attributes."""
+        return {
+            ATTR_OPTIONS: self.options,
+        }
+
+    @property
+    @final
+    def state(self) -> str | None:
+        """Return the entity state."""
+        if self.current_option is None or self.current_option not in self.options:
+            return None
+        return self.current_option
+
+    @property
+    def options(self) -> list[str]:
+        """Return a set of selectable options."""
+        return self._attr_options
+
+    @property
+    def current_option(self) -> str | None:
+        """Return the selected entity option to represent the entity state."""
+        return self._attr_current_option
+
+    def select_option(self, option: str | None) -> None:
```

Comment:
Does this look right?
```python
    def select_option(self, option: str | None) -> None:
        """Change the selected option."""
        if option not in self._attr_options:
            options = ", ".join(self._attr_options)
            _LOGGER.warning(
                f"Invalid option: {option} (possible options: {options})",
            )
            return
        self._attr_current_option = option
        self.async_write_ha_state()
```

Reply:
I have this in a local branch, but I'm not sure if it makes sense to PR against your initial PR 😅

Reply:
The implementation of changing an option is implementation done by the integration, not by the entity component.

For example, if one selects an option, an action usually comes out of that. E.g., in this example, changing the actual speed of the spaceship. A result of that might be that the state has changed, but that is not up to the entity component, but the integration implementation.

Reply:
Ah, so an integration might (or might not) expose a service to change the option, and then update the `select` accordingly?

Conversely, simply selecting a different option in the UI, should tell the integration to do an action... am I understanding that right?

Reply:
> Ah, so an integration might (or might not) expose a service to change the option

An integration must implement the method to change the option (otherwise a sensor entity should be used).

> Conversely, simply selecting a different option in the UI, should tell the integration to do an action

Well it tells the integration that the user wants to change something :) For example, in case of a light: change state of the light to on (turn it on is most likely the result in that case).